File: typescript/systemdesign/topKFeedMerger.ts
/*
Merges k user timelines, each already sorted most-recent-first, into a
single feed ordered by recency, using a min-of-k... actually max-of-k heap
over the current head of each timeline so the overall merge runs in
O(n log k) instead of concatenating and sorting everything.

status - completed
*/

import { logp } from "../utils/logger";

interface Post {
  authorId: string;
  postedAtMs: number;
  text: string;
}

interface HeapEntry {
  post: Post;
  timelineIndex: number;
  postIndex: number;
}

class MaxHeap<T> {
  private readonly items: T[] = [];

  constructor(private readonly compare: (a: T, b: T) => number) {}

  get size(): number {
    return this.items.length;
  }

  push(item: T): void {
    this.items.push(item);
    let i = this.items.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.compare(this.items[i], this.items[parent]) <= 0) break;
      [this.items[i], this.items[parent]] = [this.items[parent], this.items[i]];
      i = parent;
    }
  }

  pop(): T | undefined {
    if (this.items.length === 0) return undefined;
    const top = this.items[0];
    const last = this.items.pop()!;
    if (this.items.length > 0) {
      this.items[0] = last;
      let i = 0;
      const n = this.items.length;
      while (true) {
        const left = 2 * i + 1;
        const right = 2 * i + 2;
        let largest = i;
        if (left < n && this.compare(this.items[left], this.items[largest]) > 0) largest = left;
        if (right < n && this.compare(this.items[right], this.items[largest]) > 0) largest = right;
        if (largest === i) break;
        [this.items[i], this.items[largest]] = [this.items[largest], this.items[i]];
        i = largest;
      }
    }
    return top;
  }
}

// merges k already-sorted (most-recent-first) timelines into one feed
function mergeTimelines(timelines: Post[][], limit: number = Infinity): Post[] {
  const heap = new MaxHeap<HeapEntry>((a, b) => a.post.postedAtMs - b.post.postedAtMs);

  timelines.forEach((timeline, timelineIndex) => {
    if (timeline.length > 0) {
      heap.push({ post: timeline[0], timelineIndex, postIndex: 0 });
    }
  });

  const merged: Post[] = [];
  while (heap.size > 0 && merged.length < limit) {
    const entry = heap.pop()!;
    merged.push(entry.post);

    const nextIndex = entry.postIndex + 1;
    const timeline = timelines[entry.timelineIndex];
    if (nextIndex < timeline.length) {
      heap.push({ post: timeline[nextIndex], timelineIndex: entry.timelineIndex, postIndex: nextIndex });
    }
  }
  return merged;
}

// usage scenarios

const alice: Post[] = [
  { authorId: "alice", postedAtMs: 900, text: "alice post 3" },
  { authorId: "alice", postedAtMs: 500, text: "alice post 2" },
  { authorId: "alice", postedAtMs: 100, text: "alice post 1" },
];
const bob: Post[] = [
  { authorId: "bob", postedAtMs: 800, text: "bob post 2" },
  { authorId: "bob", postedAtMs: 200, text: "bob post 1" },
];
const carol: Post[] = [{ authorId: "carol", postedAtMs: 1000, text: "carol post 1" }];

const feed = mergeTimelines([alice, bob, carol]);
logp("merged feed, most recent first:");
feed.forEach((post) => logp(`  t=${post.postedAtMs} [${post.authorId}] ${post.text}`));

const timestamps = feed.map((p) => p.postedAtMs);
const isDescending = timestamps.every((t, i) => i === 0 || timestamps[i - 1] >= t);
logp(`feed is correctly sorted most-recent-first: ${isDescending}`);
logp(`total posts merged: ${feed.length} (expected 6)`);

const top3 = mergeTimelines([alice, bob, carol], 3);
logp(`top 3 most recent across all users: ${top3.map((p) => `${p.authorId}@${p.postedAtMs}`).join(", ")}`);

File: python/systemdesign/top_k_feed_merger.py
"""
Merge k user timelines, each already sorted by time descending, into
a single feed ordered most-recent-first using a heap-based k-way
merge instead of concatenating and re-sorting everything.

status - completed
"""

import heapq

from base_logger.logging_event import create_logger


class TopKFeedMerger:

    log = create_logger(__name__)

    def merge_feeds(self, timelines, limit=None):
        heap = []
        for feed_index, timeline in enumerate(timelines):
            if timeline:
                post = timeline[0]
                heapq.heappush(heap, (-post["timestamp"], feed_index, 0, post))

        merged = []
        while heap and (limit is None or len(merged) < limit):
            neg_ts, feed_index, item_index, post = heapq.heappop(heap)
            merged.append(post)

            next_index = item_index + 1
            timeline = timelines[feed_index]
            if next_index < len(timeline):
                next_post = timeline[next_index]
                heapq.heappush(heap, (-next_post["timestamp"], feed_index, next_index, next_post))

        return merged


if __name__ == "__main__":
    log = create_logger("top_k_feed_merger_demo")

    alice_timeline = [
        {"user": "alice", "timestamp": 100, "text": "alice post 3"},
        {"user": "alice", "timestamp": 60, "text": "alice post 2"},
        {"user": "alice", "timestamp": 10, "text": "alice post 1"},
    ]
    bob_timeline = [
        {"user": "bob", "timestamp": 90, "text": "bob post 2"},
        {"user": "bob", "timestamp": 20, "text": "bob post 1"},
    ]
    carol_timeline = [
        {"user": "carol", "timestamp": 150, "text": "carol post 1"},
    ]

    merger = TopKFeedMerger()
    merged_feed = merger.merge_feeds([alice_timeline, bob_timeline, carol_timeline])

    timestamps = [post["timestamp"] for post in merged_feed]
    log.info("merged feed timestamps: %s", timestamps)
    log.info("sorted descending: %s", timestamps == sorted(timestamps, reverse=True))

    top_3 = merger.merge_feeds([alice_timeline, bob_timeline, carol_timeline], limit=3)
    log.info("top 3 most recent posts: %s", [(p["user"], p["timestamp"]) for p in top_3])
